src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventDeserializer.java

```diff
@@ -240,6 +242,27 @@ public EventDataDeserializer getEventDataDeserializer(EventType eventType) {
         return eventDataDeserializer != null ? eventDataDeserializer : defaultEventDataDeserializer;
     }
 
+    public EventDeserializer detectChecksumType(ByteArrayInputStream inputStream)
+        throws IOException {
+        inputStream.mark(EventHeaderV4.HEADER_LEN);
+        EventHeader eventHeader = eventHeaderDeserializer.deserialize(inputStream);
+        inputStream.reset();
+
+        inputStream.mark((int) (EventHeaderV4.HEADER_LEN + eventHeader.getDataLength()));
+        inputStream.skip(EventHeaderV4.HEADER_LEN); // skip header, not interpret it twice
+        FormatDescriptionEventDataDeserializer deserializer = new FormatDescriptionEventDataDeserializer();
+        FormatDescriptionEventData formatDescriptionEventData = deserializer.deserialize(inputStream);
+        int eventLength = formatDescriptionEventData.getEventLength();
+        // 1 byte means checksum algo, 4 byte is crc checksum content
+        if (eventLength + 1 + 4 == eventHeader.getDataLength()) {
```

Comment:
can all of this logic be done "live" as part of [EventDeserializer::nextEvent](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventDeserializer.java#L192)?

Reply:
I don't think it is a good idea to do it in `nextEvent`. We only need to detect once in a single file and check the first event of binlog. So it seems no need to add extra flag to check every time in `nextEvent`.

Reply:
The reason I asked is https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L868. If EventDeserializer can determine the presence of the checksum on its own - we can avoid BinaryLogClient mutating the state of EventDeserializer (something that would have to be done 1.0.0 anyway where all API is immutable). This would also remove the need in https://github.com/shyiko/mysql-binlog-connector-java/pull/219/files#diff-8ae42522a3e6dce8339dccb9f4a701a3R61 (which can have adverse perf implications depending on what inputStream really is; however slight) and https://github.com/shyiko/mysql-binlog-connector-java/pull/219/files#diff-a7680db73010e96a8f00b8f2b6229cbeR223.

To be clear: I'm OK with merging the PR as is. Just trying to weigh the options here.

Reply:
As far as I used your library, to avoid setting checksum using `show global variables like 'binlog_checksum'` seems impossible because I will remember where I leave last time and restart from the middle of binlog stream and no `FormatDescriptionEvent`. 

Making it immutable is a good idea, but will only works for `BinaryLogFileReader` as can see.

Reply:
Ok, I see. One more question: it should be possible that different binlog file from same server uses different checksum algo, right? So, if we are going to do live interpretation of checksum algo, we need to handle all `FORMAT_DESCRIPTION` rather than just the first, right?